helix-term/src/ui/editor.rs

```diff
@@ -559,19 +563,72 @@ impl EditorView {
         );
     }
 
-    fn insert_mode(&self, cx: &mut commands::Context, event: KeyEvent) {
-        if let Some(command) = self.keymaps[&Mode::Insert].get(&event) {
-            command.execute(cx);
-        } else if let KeyEvent {
-            code: KeyCode::Char(ch),
-            ..
-        } = event
-        {
-            commands::insert::insert_char(cx, ch);
+    /// Handle events by looking them up in self.keymaps. Note that the event
+    /// may be at the top level or any nested subkeymap inside keymaps. Because of
+    /// this, handling an event does not necessarily mean that a command was
+    /// executed; it may be a multi keystroke mapping. Returns None if event was
+    /// handled, Some(n) otherwise where n denotes number of pending events to
+    /// be handled
+    fn handle_keymap_event(
+        &mut self,
+        mode: Mode,
+        cxt: &mut commands::Context,
+        event: KeyEvent,
+    ) -> Option<usize> {
+        let keymap = match self.current_subkeymap {
+            Some(ref keymap) => keymap,
+            None => &self.keymaps[&mode],
+        };
+        if let Some(keynode) = keymap.get(&event) {
+            match keynode {
+                KeyNode::KeyCommand(command) => {
+                    command.execute(cxt);
+                    self.current_subkeymap = None;
+                    self.subkeymap_level = 0;
+                }
+                KeyNode::SubKeymap(subkeymap) => {
+                    self.current_subkeymap = Some(subkeymap.clone());
+                    self.subkeymap_level += 1;
+                }
```

Comment:
I don't think this is a good way to solve this. Ideally we want to be able to show infobox as well since it requires similar thing.

Reply:
Since we can now specify multi key mappings, I was thinking of moving all of the builtin ones like `gy`, `gr`, `<space>b`, etc from the `mode_info` macros to `keymap.rs` and then render the infobox from here (in the `KeyNode::SubKeymap(subkeymap) =>` branch) instead of doing it from the macro generated mode functions. I think it'll solve the problem since the default keys shown in infobox after this change aren't a special case and are essentially defined the same way as user defined multi key remappings.